src/Agent.Plugins/PipelineArtifact/FileShareProvider.cs

```diff
@@ -0,0 +1,104 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using System.Threading.Tasks.Dataflow;
+using Agent.Sdk;
+using Microsoft.TeamFoundation.Build.WebApi;
+using Microsoft.VisualStudio.Services.BlobStore.Common;
+using Microsoft.VisualStudio.Services.Content.Common.Tracing;
+using Microsoft.VisualStudio.Services.Agent.Util;
+using Microsoft.VisualStudio.Services.Content.Common;
+
+
+namespace Agent.Plugins.PipelineArtifact
+{
+    internal class FileShareProvider: IArtifactProvider
+    {
+        private readonly AgentTaskPluginExecutionContext context;
+        private readonly CallbackAppTraceSource tracer;
+        private const int defaultParallelCount = 1;
+
+        public FileShareProvider(AgentTaskPluginExecutionContext context, CallbackAppTraceSource tracer)
+        {
+            this.context = context;
+            this.tracer = tracer;
+        }
+
+        public async Task DownloadSingleArtifactAsync(PipelineArtifactDownloadParameters downloadParameters, BuildArtifact buildArtifact, CancellationToken cancellationToken)
+        {
+            var downloadRootPath = buildArtifact.Resource.Data + Path.DirectorySeparatorChar + buildArtifact.Name;
+            await this.CopyFileShare(downloadParameters.ProjectId, downloadRootPath, downloadParameters.TargetDirectory, downloadParameters.MinimatchFilters, cancellationToken);
+        }
+
+        public async Task DownloadMultipleArtifactsAsync(PipelineArtifactDownloadParameters downloadParameters, IEnumerable<BuildArtifact> buildArtifacts, CancellationToken cancellationToken)
+        {
+            foreach (var buildArtifact in buildArtifacts)
+            {
+                var dirPath = Path.Combine(downloadParameters.TargetDirectory, buildArtifact.Name);
+                var downloadRootPath = buildArtifact.Resource.Data;
+                await this.CopyFileShare(downloadParameters.ProjectId, downloadRootPath, dirPath, downloadParameters.MinimatchFilters, cancellationToken, isSingleArtifactDownload: false);
+            }
+        }
+
+        public async Task PublishArtifactAsync(string sourcePath, string destPath, int parallelCount)
+        {
+            await this.DirectoryCopyWithMiniMatch(sourcePath, destPath, CancellationToken.None, parallelCount);
```

Comment:
CancellationToken should really be passed by the caller. The agent has a cancellation token from the start of the plugini and should be plumbed through here. A good rule of thumb is to have cancellation tokens passed in as part of asynchronous functions in C#:
https://docs.microsoft.com/en-us/dotnet/standard/asynchronous-programming-patterns/task-based-asynchronous-pattern-tap#cancellation-optional